packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SurroundingPairScopeHandler/SurroundingPairScopeHandler.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  Direction,
+  Position,
+  SurroundingPairScopeType,
+  TextEditor,
+  showError,
+  type ScopeType,
+} from "@cursorless/common";
+import type { LanguageDefinitions } from "../../../../languages/LanguageDefinitions";
+import { BaseScopeHandler } from "../BaseScopeHandler";
+import { compareTargetScopes } from "../compareTargetScopes";
+import { TargetScope } from "../scope.types";
+import { ScopeIteratorRequirements } from "../scopeHandler.types";
+import { getDelimiterOccurrences } from "./getDelimiterOccurrences";
+import { getDelimiterRegex } from "./getDelimiterRegex";
+import { getIndividualDelimiters } from "./getIndividualDelimiters";
+import { getSurroundingPairOccurrences } from "./getSurroundingPairOccurrences";
+import { ide } from "../../../../singletons/ide.singleton";
+import { createTargetScope } from "./createTargetScope";
+import { stronglyContains } from "./stronglyContains";
+
+export class SurroundingPairScopeHandler extends BaseScopeHandler {
+  public readonly iterationScopeType;
+  protected isHierarchical = true;
+
+  constructor(
+    private languageDefinitions: LanguageDefinitions,
+    public readonly scopeType: SurroundingPairScopeType,
+    private languageId: string,
+  ) {
+    super();
+    const iterationScopeType: ScopeType = {
+      type: "oneOf",
+      scopeTypes: [this.scopeType, { type: "document" }],
+    };
+    this.iterationScopeType = iterationScopeType;
+  }
+
+  *generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+    hints: ScopeIteratorRequirements,
+  ): Iterable<TargetScope> {
+    const individualDelimiters = getIndividualDelimiters(
+      this.scopeType.delimiter,
+      this.languageId,
+    );
+    const delimiterRegex = getDelimiterRegex(individualDelimiters);
+
+    if (this.scopeType.forceDirection != null) {
+      //  DEPRECATED @ 2024-07-01
+      void showError(
+        ide().messages,
+        "deprecatedForceDirection",
+        "forceDirection not supported. Use 'next pair' or 'previous pair' instead",
+      );
+      return;
+    }
+
+    const delimiterOccurrences = getDelimiterOccurrences(
+      this.languageDefinitions.get(this.languageId),
+      editor.document,
+      individualDelimiters,
+      delimiterRegex,
+    );
+
+    const surroundingPairs = getSurroundingPairOccurrences(
+      individualDelimiters,
+      delimiterOccurrences,
+    );
```

Comment:
This is also quite brute-force. So if I'm sitting next to a paren and say "take round", cursorless regexes the entire file, then iterates over all delimiters to match them up

I don't want to prematurely optimise, but elsewhere we've been quite careful to leverage the fact we have generators so that we can support big log files. I'm not sure why we don't care when it comes to pairs?

Reply:
I wanted to do the same thing as we do with the syntax scopes using Tree sitter. The position of each scope is defined in the document and your cursor position has no bearing on it.